python/src/spark_rapids_ml/dbscan.py

```diff
@@ -0,0 +1,747 @@
+        Dict[str, Any],
+    ]:
+        import cupy as cp
+        import cupyx
+
+        dtype = self.dtype
+        n_cols = self.n_cols
+        array_order = self._fit_array_order()
+        pred_name = self._get_prediction_name()
+        idCol_name = self.getIdCol()
+        logger = get_logger(self.__class__)
+
+        cuda_managed_mem_enabled = (
+            _get_spark_session().conf.get("spark.rapids.ml.uvm.enabled", "false")
+            == "true"
+        )
+
+        inputs = []  # type: ignore
+
+        idCol = list(
+            self.idCols_[0].value
+            if len(self.idCols_) == 1
+            else np.concatenate([chunk.value for chunk in self.idCols_])
+        )
+
+        for pdf_bc in self.raw_data_:
+            pdf = pd.DataFrame(data=pdf_bc.value, columns=self.processed_input_cols)
+
+            if self.multi_col_names:
+                features = np.array(pdf[self.multi_col_names], order=array_order)
+            elif self.use_sparse_array:
+                # sparse vector
+                features = _read_csr_matrix_from_unwrapped_spark_vec(pdf)
+            else:
+                # dense vector
+                features = np.array(list(pdf[alias.data]), order=array_order)
+
+            # experiments indicate it is faster to convert to numpy array and then to cupy array than directly
+            # invoking cupy array on the list
+            if cuda_managed_mem_enabled:
+                features = (
+                    cp.array(features)
+                    if self.use_sparse_array is False
+                    else cupyx.scipy.sparse.csr_matrix(features)
+                )
+
+            inputs.append(features)
+
+        if isinstance(inputs[0], pd.DataFrame):
+            concated = pd.concat(inputs)
+        else:
+            # features are either cp or np arrays here
+            concated = _concat_and_free(inputs, order=array_order)
+
+        def _cuml_fit(
+            dfs: FitInputType,
+            params: Dict[str, Any],
+        ) -> Dict[str, Any]:
+            from cuml.cluster.dbscan_mg import DBSCANMG as CumlDBSCANMG
+            from pyspark import BarrierTaskContext
+
+            context = BarrierTaskContext.get()
+            partition_id = context.partitionId()
+
+            logger = get_logger(self.__class__)
+
+            dbscan = CumlDBSCANMG(
+                handle=params[param_alias.handle],
+                output_type="cudf",
+                eps=self.getOrDefault("eps"),
+                min_samples=self.getOrDefault("min_samples"),
+                metric=self.getOrDefault("metric"),
+                max_mbytes_per_batch=self.getOrDefault("max_mbytes_per_batch"),
+                calc_core_sample_indices=self.getOrDefault("calc_core_sample_indices"),
+                verbose=self.verbose,
+            )
+            dbscan.n_cols = params[param_alias.num_cols]
+            dbscan.dtype = np.dtype(dtype)
+
+            res = list(dbscan.fit_predict(concated).to_numpy())
```

Comment:
Are we able to construct concated from dfs? If so, we can avoid the duplicating code snippet that construct concated from the dataframe.